packages/jni-swig-stub/src/main/jni/realm_api_helpers.cpp

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "realm_api_helpers.h"
+#include <vector>
+
+using namespace realm::jni_util;
+
+realm_notification_token_t *
+register_results_notification_cb(realm_results_t *results, jobject callback) {
+    auto jenv = get_env();
+    static jclass notification_class = jenv->FindClass("io/realm/interop/NotificationCallback");
+    static jmethodID on_change_method = jenv->GetMethodID(notification_class, "onChange", "(J)V");
+
+    return realm_results_add_notification_callback(
+            results,
+            // Use the callback as user data
+            static_cast<jobject>(get_env()->NewGlobalRef(callback)),
+            [](void *userdata) {
+                get_env(true)->DeleteGlobalRef(static_cast<jobject>(userdata));
+            },
+            // change callback
+            [](void *userdata, const realm_collection_changes_t *changes) {
```

Comment:
This callback is shared between `list` and `results`, would be nice if it was only defined once. But can just add a FIXME for now

Reply:
Added a fix referencing this. Probably fits will into implementing _set_ and _dict_ change listeners ;)